### ISSUE TITLE
Dependency not found klona/full

### ISSUE THREAD

**Issue by provokcx** [nuxt-modules/i18n#1004]
### Version
nuxt-i18n: 6.16.0
nuxt: 2.14.12

#### [mode](https://nuxtjs.org/guides/configuration-glossary/configuration-mode): 
 - [X] universal
 - [ ] spa

### Steps to reproduce
Upgrade or install nuxt-i18n.

### What is Expected?
**The lib is expected to work**.

### What is actually happening?
**It doesn't**... (See the issue's title)

```
This dependency was not found:                                                                
* klona/full in ./.nuxt/nuxt-i18n/plugin.main.js
To install it, you can run: npm install --save klona/full
```

**Comment by rchl** [comment]
https://github.com/nuxt-community/i18n-module/issues/994

**Comment by provokcx** [comment]
> #994

Already tried it.
Did not work.
It has nothing to do with caching.

**Comment by divine** [comment]
Hello,

Well, if version 6.16.0 was installed then klona would be here, however, it's not found.

Basically could you please run yarn install or npm install? Clearing .nuxt cache might help also.

Thanks!

**Comment by provokcx** [comment]
- [X] `yarn remove nuxt-i18n`
- [X] `yarn add -S nuxt-i18n`
- [X] Removed `./node_modules`
- [X] Removed `./cache`
- [X] Removed `./.nuxt`
- [X] `yarn install`

Still not working.
`klona` is present in  `./node_modules`... but without the _"full"_ directory within it.

Here's klona package.json's exports section.
![Capture d’écran 2020-12-22 à 00 17 54](https://user-images.githubusercontent.com/37068300/102831104-3c1f6c80-43eb-11eb-839d-ad9f63ff0af5.png)

They reference the export of the _"full"_ directory which contains a js file needed to run nuxt-i18n (since you import that js file in `./.nuxt/nuxt-i18n/plugin.main.js`).

The files `klona/full/index.mjs` and `klona/full/index.js` are not there in the v6.16.0 of nuxt-i18n, thus it doesn't work.

**Comment by rchl** [comment]
Check this: https://unpkg.com/browse/klona@2.0.4/

There is something wrong with your system/setup. Possibly with the yarn/npm cache.

**Comment by divine** [comment]
I've just spinned up new nuxt app and I can't recreate the problem that you're having.

**Comment by provokcx** [comment]
I think that I have a clue.

Inside `./node_modules/nuxt-i18n` there is klona@2.0.4.
There is klona@1.1.2 at `./node_modules/klona`.

Since the file importing klona is `./.nuxt/nuxt-i18n/plugin.main.js`, I think that it searches for _"full"_ inside `./node_modules/klona@1.1.2` and not `./node_modules/nuxt-i18n/klona@2.0.4`.

Would that be possible?
Wouldn't peerDependencies solve the whole thing?

**Comment by divine** [comment]
> I think that I have a clue.
> 
> Inside `./node_modules/nuxt-i18n` there is klona@2.0.4.
> There is klona@1.1.2 at `./node_modules/klona`.
> 
> Since the file importing klona is `./.nuxt/nuxt-i18n/plugin.main.js`, I think that it searches for _"full"_ inside `./node_modules/klona@1.1.2` and not `./node_modules/nuxt-i18n/klona@2.0.4`.
> 
> Would that be possible?
> Wouldn't peerDependencies solve the whole thing?

I have just reproduced it locally.  I've installed `klona@1.1.2` and after added nuxt-i18n. It throws errors on compile because klona is an older version which doesn't contain full directory https://github.com/lukeed/klona/releases/tag/v2.0.0

I've moved klona to **peerDependencies** as It's just throwing warning but not installing newer version:

```js
warning " > nuxt-i18n@6.16.0" has unmet peer dependency "klona@^2.0.4".
```

@rchl what do you think? Should we move to **peerDependencies** or just keep it as it is.

Basically, it doesn't resolve the issue, @Alphability  you need to upgrade your local `klona` version to ^2.

Thanks!

**Comment by provokcx** [comment]
Since some of nuxt-i18n code runs outside of the package itself I recommend you to install the dependencies **needed on the outside as peerDependencies**.
That's what they're for if you read [yarn](https://classic.yarnpkg.com/en/docs/dependency-types/#toc-peerdependencies) or npm documentation.

Those dependencies are **not auto-installed**.
So, you'll need to tell your users to install them.

Your setup section should just look like this:
```
# yarn
yarn add nuxt-i18n klona@^2.0.4

# npm
npm i nuxt-i18n klona@^2.0.4
```

**Comment by rchl** [comment]
It does feel like it would make sense for modules to specify dependencies that they use in plugins as peer dependencies. It's not a normal practice though, as far as I can tell, and it would be kinda unusual to do that.

Not to mention that requiring the user to manage dependencies himself would make things a lot less user-friendly and potentially error-prone. For example, a module can enforce certain dependency versions while it's not as straight-forward with peer dependencies. It can still specify which version should be used but if the user doesn't match that, it will only result in an easy-to-miss warning when installing dependencies.

As for this particular problem, you can of course just install `klona` dependency manually (without having it in peer dependencies), and then that version will be used and things will work for `nuxt-i18n`. Only that you risk problems with that other code that requires klona 1.x (especially if it's used from a plugin of another module). And that problem would apply even if `klona` would be specified in `peerDependencies`. So there is no perfect solution really.

**Comment by rchl** [comment]
@pi0 @Atinux This is an interesting problem and I wonder if you have any take on it.

**Comment by provokcx** [comment]
> Not to mention that requiring the user to manage dependencies himself would make things a lot less user-friendly and potentially error-prone. For example, a module can enforce certain dependency versions while it's not as straight-forward with peer dependencies.

I totally understand your point.

From my perspective the real less user friendly thing and error prone to do is to let open the possibility to get an error like this with complex projects.

The more dependencies you have and/or need the more likely you are to get this error... and the more likely to waste a lot of time trying to understand where it comes from.

Yes peerDependencies are scary... but they're here for a reason.
They're not the problem, they're here to solve this particular one.

> It can still specify which version should be used but if the user doesn't match that, it will only result in an easy-to-miss warning when installing dependencies.

In my opinion, It is never a bad idea to make people accountable (especially developers) for what they do.

@Atinux @pi0 what is your opinion about this pretty interesting matter?

**Comment by rchl** [comment]
`peerDependencies` doesn't fix the problem in case two modules require the same dependency but in different, incompatible versions. The user can only install a single version.

**Comment by provokcx** [comment]
Yeah that is a problem... but this problem comes directly from the fact that we can add plugins with a module.

Thanks for the chat @rchl, this is really interesting! 🙏 

I am eager to read from Atinux and pi0 on this!

**Comment by pi0** [comment]
Hi @rchl. You can resolve full path of klona on behalf of module for exact version that requires: 


```js
nuxt.options.alias['klona'] = require.resolve('klona/lite')
// Alt: nuxt.options.alias['~i18n-klona'] = require.resolve('klona/lite')
```

For a longer answer, I'm compiling runtime instead of emitting as plugin template for image and auth modules and adding this pattern with new module template and nuxt/kit. This resolves implicit dependency issues as well as making runtime lintable/typed.

**Comment by ABleas** [comment]
I found this issue when doing an upgrade on dependencies for my project.

I could install `klona/full` as a peerDependency as suggested, but

- I don't want to know what it is, nor do I much care.
- The error suggests this is a dependency of nuxt-i18n, why do I have to do extra steps to make this package work?

Just from a naive, dumb user point of view, should this not be resolved on the i18n module side of things?

**Comment by provokcx** [comment]
> * I don't want to know what it is, nor do I much care.

@ABleas 
**THE WORLD IS BURNING BECAUSE OF PEOPLE THAT DON'T WANT TO KNOW OR DON'T REALLY CARE** 👏 👏 👏 

Just because a 404 is not as lethal as a bridge collapsing doesn't mean that you're free to stop learning how to do your job 😘

**Comment by ABleas** [comment]
> Just because a 404 is not as lethal as a bridge collapsing doesn't mean that you're free to stop learning how to do your job 😘

Sorry for sounding gruff. I appreciate that it's important to somebody, but it isn't my job to understand everything that goes on under the hood of this wonderful module. That would make my job impossible!

I should also say, I am really happy with nuxt-i18n. It is a pleasure to use.

**Comment by daishizenSensei** [comment]
Hello, it seems that this issue still be here. We've update to the last release and have this error : 
Failed to compile with 1 errors
This dependency was not found: 
* ~i18n-klona in ./.nuxt/nuxt-i18n/plugin.main.js
To install it, you can run: npm install --save ~i18n-klona

**Comment by daishizenSensei** [comment]
Of course npm install --save ~i18n-klona doesn't work

**Comment by rchl** [comment]
Delete node_modules and try again:
```
rm -rf node_modules
npm i
```

**Comment by daishizenSensei** [comment]
It's working, thank you :)
